migration_challenge_keras_image/Local Notebook.ipynb

```diff
@@ -138,39 +138,67 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "import os\n"
-   ]
-  },
-  {
-   "cell_type": "code",
-   "execution_count": null,
-   "metadata": {},
-   "outputs": [],
-   "source": [
-    "%%time\n",
+    "%time\n",
+    "\n",
     "!rm -rf data/train\n",
     "!rm -rf data/test\n",
+    "import os\n",
+    "import concurrent.futures\n",
+    "from random import randint\n",
+    "\n",
+    "[os.makedirs(os.path.join(\"data/train\", \"digit-%d\" % ix), exist_ok=True) for ix in range(10)]\n",
+    "[os.makedirs(os.path.join(\"data/test\", \"digit-%d\" % ix), exist_ok=True) for ix in range(10)] \n",
+    "\n",
     "\n",
     "# This can take a while due to the number of images\n",
     "def save_to_disk(x, y, base_folder):\n",
     "    \"\"\"Save an image classification dataset to disk as JPEGs in label-named folders\"\"\"\n",
     "    for ix in range(len(y)):\n",
-    "        label_str = \"digit-%d\" % y[ix]\n",
-    "        os.makedirs(os.path.join(base_folder, label_str), exist_ok=True)\n",
+    "        label_dir = \"digit-%d\" % y[ix]\n",
+    "        label_str = \"digit-%d-%d\" % (randint(0, 100000), y[ix])\n",
     "        tf.keras.preprocessing.image.save_img(\n",
-    "            os.path.join(base_folder, label_str, \"%s-%06d.jpg\" % (label_str, ix)),\n",
+    "            os.path.join(base_folder, label_dir, \"%s-%06d.jpg\" % (label_str, ix)),\n",
     "            # This function expects a channels dimension, which we'll add in first:\n",
     "            np.expand_dims(x[ix], 0),\n",
     "            data_format=\"channels_first\"\n",
     "        )\n",
-    "\n",
-    "print(\"Saving training data...\")\n",
+    "    return \"Saving...\"\n",
+    "        \n",
+    "print(\"Saving training and test data...\")\n",
     "os.makedirs(\"data/train\", exist_ok=True)\n",
-    "save_to_disk(x_train_raw, y_train_raw, \"data/train\")\n",
-    "print(\"Saving test data...\")\n",
     "os.makedirs(\"data/test\", exist_ok=True)\n",
-    "save_to_disk(x_test_raw, y_test_raw, \"data/test\")\n",
-    "print(\"Done!\")\n"
+    "\n",
+    "# Splitting data to save in thread for faster speed\n",
+    "thread_split = 5\n",
+    "x = np.array_split(x_train_raw, thread_split)\n",
+    "y = np.array_split(y_train_raw, thread_split)\n",
+    "\n",
+    "\n",
+    "with concurrent.futures.ProcessPoolExecutor() as executor:\n",
+    "    results = [executor.submit(save_to_disk,x[i], y[i],\"data/train\") for i in range(thread_split)]\n",
+    "    executor.submit(save_to_disk,x_test_raw, y_test_raw,\"data/test\")\n",
+    "    for f in concurrent.futures.as_completed(results):\n",
+    "        print(f.result())\n",
```

Comment:
Not a big fan of returning `Saving...` when `save_to_disk()` completes as it's a weird function contract.

How about we just continue returning nothing, and instead keep a counter of how many of our jobs have finished? Like:

```python
n_done = 0
...
# (Set up the results list)
...
for f in concurrent.futures.as_completed(results):
    n_done += 1
    print(f"{n_done} of {len(results)} jobs finished")
```

Reply:
Updated progress logging in 59fc64ff88a367f472b73fe52b3e43622bcf4c26